lib/reqIdGenFactory.js

```diff
@@ -1,6 +1,11 @@
 'use strict'
 
-module.exports = function (requestIdHeader) {
+module.exports = function (requestIdHeader, userGenReqIdFunc) {
+  if (userGenReqIdFunc) {
+    return function genReqId (req) {
+      return req.headers[requestIdHeader] || userGenReqIdFunc(req)
+    }
+  }
```

Comment:
I would avoid allocating this additional function here, and move the check on `requestIdHeader` to https://github.com/fastify/fastify/blob/77abd501022ccf3e1a49f84e97e89a3648af3dca/fastify.js#L275. I think the change would be more deliberate in that way.

In turn, I would remove the `||` in the `genReqId` function in here.